solver/src/main/java/org/chocosolver/solver/variables/impl/BitsetIntVarImpl.java

```diff
@@ -191,30 +191,35 @@ public boolean removeValue(int value, ICause cause) throws ContradictionExceptio
     @Override
     public boolean removeValues(IntIterableSet values, ICause cause) throws ContradictionException {
         assert cause != null;
-        int olb = getLB();
-        int oub = getUB();
-        int nlb = values.nextValue(olb - 1);
-        int nub = values.previousValue(oub + 1);
-        if (nlb > oub || nub < olb) {
-            return false;
-        }
-        int i;
-        // look for the new lb
-        while (nlb == olb && olb < Integer.MAX_VALUE) {
-            i = VALUES.nextSetBit(nlb + 1 - OFFSET);
-            olb = i > -1 ? i + OFFSET : Integer.MAX_VALUE;
+        boolean hasChanged = false, fixpoint;
+        int nlb, nub;
+        do {
+            int olb = getLB();
+            int oub = getUB();
             nlb = values.nextValue(olb - 1);
-        }
-        if (nlb <= nub) {
-            // look for the new ub
-            while (nub == oub && oub > Integer.MIN_VALUE) {
-                i = VALUES.prevSetBit(nub - 1 - OFFSET);
-                oub = i > -1 ? i + OFFSET : Integer.MIN_VALUE;
-                nub = values.previousValue(oub + 1);
+            nub = values.previousValue(oub + 1);
+            if (!hasChanged && (nlb > oub || nub < olb)) {
+                return false;
             }
-        }
-        // the new bounds are now known, delegate to the right method
-        boolean hasChanged = updateBounds(olb, oub, cause);
+            int i;
+            // look for the new lb
+            while (nlb == olb && olb < Integer.MAX_VALUE) {
+                i = VALUES.nextSetBit(nlb + 1 - OFFSET);
+                olb = i > -1 ? i + OFFSET : Integer.MAX_VALUE;
+                nlb = values.nextValue(olb - 1);
+            }
+            if (nlb <= nub) {
+                // look for the new ub
+                while (nub == oub && oub > Integer.MIN_VALUE) {
+                    i = VALUES.prevSetBit(nub - 1 - OFFSET);
+                    oub = i > -1 ? i + OFFSET : Integer.MIN_VALUE;
+                    nub = values.previousValue(oub + 1);
+                }
+            }
+            // the new bounds are now known, delegate to the right method
+            fixpoint = updateBounds(olb, oub, cause);
```

Comment:
I think that it comes from finishing conditions of the algorithm, but we are using "old" bounds variables her instead of new ones (nlb and nub). It might be clearer to call them.